store/secondary.go

```diff
@@ -0,0 +1,183 @@
+package store
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type ISecondary interface {
+	Fallbacks() []PrecomputedKeyStore
+	Caches() []PrecomputedKeyStore
+	Enabled() bool
+	Ingress() chan<- PutNotif
+	CachingEnabled() bool
+	FallbackEnabled() bool
+	HandleRedundantWrites(ctx context.Context, commitment []byte, value []byte) error
+	MultiSourceRead(context.Context, []byte, bool, func([]byte, []byte) error) ([]byte, error)
+	StreamProcess(context.Context)
+}
+
+type PutNotif struct {
+	Commitment []byte
+	Value      []byte
+}
+
+// SecondaryRouter ... routing abstraction for secondary storage backends
+type SecondaryRouter struct {
+	stream chan PutNotif
+	log    log.Logger
+
+	caches    []PrecomputedKeyStore
+	cacheLock sync.RWMutex
+
+	fallbacks    []PrecomputedKeyStore
+	fallbackLock sync.RWMutex
+}
+
+// NewSecondaryRouter ... creates a new secondary storage router
+func NewSecondaryRouter(log log.Logger, caches []PrecomputedKeyStore, fallbacks []PrecomputedKeyStore) (ISecondary, error) {
+	return &SecondaryRouter{
+		stream:    make(chan PutNotif),
+		log:       log,
+		caches:    caches,
+		cacheLock: sync.RWMutex{},
+
+		fallbacks:    fallbacks,
+		fallbackLock: sync.RWMutex{},
+	}, nil
+}
+
+func (r *SecondaryRouter) Ingress() chan<- PutNotif {
+	return r.stream
+}
+
+func (r *SecondaryRouter) Enabled() bool {
+	return r.CachingEnabled() || r.FallbackEnabled()
+}
+
+func (r *SecondaryRouter) CachingEnabled() bool {
+	r.cacheLock.RLock()
+	defer r.cacheLock.RUnlock()
+
+	return len(r.caches) > 0
+}
+
+func (r *SecondaryRouter) FallbackEnabled() bool {
+	r.fallbackLock.RLock()
+	defer r.fallbackLock.RUnlock()
+
+	return len(r.fallbacks) > 0
+}
+
+// handleRedundantWrites ... writes to both sets of backends (i.e, fallback, cache)
+// and returns an error if NONE of them succeed
+// NOTE: multi-target set writes are done at once to avoid re-invocation of the same write function at the same
+// caller step for different target sets vs. reading which is done conditionally to segment between a cached read type
+// vs a fallback read type
+func (r *SecondaryRouter) HandleRedundantWrites(ctx context.Context, commitment []byte, value []byte) error {
+	r.cacheLock.RLock()
+	r.fallbackLock.RLock()
+
+	defer func() {
+		r.cacheLock.RUnlock()
+		r.fallbackLock.RUnlock()
+	}()
+
+	sources := r.caches
+	sources = append(sources, r.fallbacks...)
+
+	key := crypto.Keccak256(commitment)
+	successes := 0
+
+	for _, src := range sources {
+		err := src.Put(ctx, key, value)
+		if err != nil {
+			r.log.Warn("Failed to write to redundant target", "backend", src.BackendType(), "err", err)
+		} else {
+			successes++
+		}
+	}
+
+	if successes == 0 {
+		return errors.New("failed to write blob to any redundant targets")
```

Comment:
should we use an errgroup here instead? If there are a bunch of sources, we should prob write to them in parallel?